src/cargo/ops/vendor.rs

```diff
@@ -383,3 +443,53 @@ fn copy_and_checksum(src_path: &Path, dst_path: &Path, buf: &mut [u8]) -> CargoR
             .with_context(|| format!("failed to write to {:?}", dst_path))?;
     }
 }
+
+fn source_id_to_dir_name(src_id: SourceId) -> String {
+    let src_type = if src_id.is_registry() {
+        "registry"
+    } else if src_id.is_git() {
+        "git"
+    } else {
+        panic!()
+    };
+    struct SourceIdShortHash(SourceId);
+    impl Hash for SourceIdShortHash {
+        fn hash<H: Hasher>(&self, state: &mut H) {
+            self.0.stable_hash(Path::new(""), state);
+        }
+    }
+    let hash = SourceIdShortHash(src_id);
+    format!("{}-{}", src_type, util::short_hash(&hash))
+}
+
+fn existing_crate_paths(
```

Comment:
Personally I'm not too worried about needing logic like this to preserve state or so between `cargo vendor` and `cargo vendor --no-merge-sources`. In my mind `cargo vendor` is the owner of the `vendor` directory and if other things are placed in there then that's for a higher level to manage rather than `cargo vendor` itself.